mantidproject/mantid
# Fix TOSCA reduction out of range error for old file

### Description of work
This PR fixes an out-of-range error when reducing an old TOSCA file 5224 on the Indirect Data Reduction interface. The problem was that the spectra that do no contain a detector are removed before performing the rebin, which would then invalidate the masked spectra indices calculated at the beginning of the algorithm.

To fix this I decoupled the masking function from the detector grouping function, and moved the masking to be before the rebinning.

Fixes #36997

### To test:
1. Open `Indirect`->`Data Reduction`
2. Select `TOSCA`
3. Load 05224
4. Click run. 
5. The reduction should be successful, without error

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
